tests/test_new_return_types.py

```diff
@@ -0,0 +1,357 @@
+"""
+Unit tests for the new return types.
+"""
+import pytest
+
+import numpy as np
+import pennylane as qml
+
+herm = np.diag([1, 2, 3, 4])
+probs_data = [
+    (None, [0]),
+    (None, [0, 1]),
+    (qml.PauliZ(0), None),
+    (qml.Hermitian(herm, wires=[1, 0]), None),
+]
+
+wires = [2, 3, 4]
+
+
+class TestSingleReturnExecute:
+    """Test that single measurements return behavior does not change."""
+
+    @pytest.mark.parametrize("wires", wires)
+    def test_state_default(self, wires):
+        """Return state with default.qubit."""
+        dev = qml.device("default.qubit", wires=wires)
+
+        def circuit(x):
+            qml.Hadamard(wires=[0])
+            qml.CRX(x, wires=[0, 1])
+            return qml.state()
+
+        qnode = qml.QNode(circuit, dev)
+        qnode.construct([0.5], {})
+        res = qml.execute_new(tapes=[qnode.tape], device=dev, gradient_fn=None)
+
+        assert res[0].shape == (2**wires,)
+        assert isinstance(res[0], np.ndarray)
+
+    @pytest.mark.parametrize("wires", wires)
+    def test_state_mixed(self, wires):
+        """Return state with default.mixed."""
+        dev = qml.device("default.mixed", wires=wires)
+
+        def circuit(x):
+            qml.Hadamard(wires=[0])
+            qml.CRX(x, wires=[0, 1])
+            return qml.state()
+
+        qnode = qml.QNode(circuit, dev)
+        qnode.construct([0.5], {})
+
+        res = qml.execute_new(tapes=[qnode.tape], device=dev, gradient_fn=None)
+
+        assert res[0].shape == (2**wires, 2**wires)
+        assert isinstance(res[0], np.ndarray)
+
+    @pytest.mark.parametrize("d_wires", wires)
+    def test_density_matrix_default(self, d_wires):
+        """Return density matrix with default.qubit."""
+        dev = qml.device("default.qubit", wires=4)
+
+        def circuit(x):
+            qml.Hadamard(wires=[0])
+            qml.CRX(x, wires=[0, 1])
+            return qml.density_matrix(wires=range(0, d_wires))
+
+        qnode = qml.QNode(circuit, dev)
+        qnode.construct([0.5], {})
+
+        res = qml.execute_new(tapes=[qnode.tape], device=dev, gradient_fn=None)
+
+        assert res[0].shape == (2**d_wires, 2**d_wires)
+        assert isinstance(res[0], np.ndarray)
+
+    @pytest.mark.parametrize("d_wires", wires)
+    def test_density_matrix_mixed(self, d_wires):
+        """Return density matrix with default.mixed."""
+        dev = qml.device("default.mixed", wires=4)
+
+        def circuit(x):
+            qml.Hadamard(wires=[0])
+            qml.CRX(x, wires=[0, 1])
+            return qml.density_matrix(wires=range(0, d_wires))
+
+        qnode = qml.QNode(circuit, dev)
+        qnode.construct([0.5], {})
+        res = qml.execute_new(tapes=[qnode.tape], device=dev, gradient_fn=None)
+        assert res[0].shape == (2**d_wires, 2**d_wires)
+        assert isinstance(res[0], np.ndarray)
+
+    def test_expval(self):
+        """Return a single expval."""
+        dev = qml.device("default.qubit", wires=2)
```

Comment:
For this test and everything below it, it might be good to test with the `default.mixed` device too (with `@pytest.mark.parametrize`).